docs/development/etcdoperatortask-design.md

```diff
@@ -0,0 +1,481 @@
+status:
+    observedGeneration: <specific observedGeneration of the resource>
+    state: <last known current state of the out-of-band task>
+    initiatedAt: <time at which task move to any other state from "pending" state>
+    lastErrors:
+    - code: <error-code>
+      description: <description of the error>
+      observedAt: <time the error was observed>
+    lastOperation:
+      description: <meaningful description>
+      state: <task state as seen at the completion of last operation>
+      lastTransitionTime: <time of transition to this state>
+
+```
+</details>
+
+## TaskHandler Interface
+
+The `TaskHandler` interface defines the contract for implementing out-of-band task logic. Each task type must provide its own handler, encapsulating admission checks, execution, and cleanup.
+
+
+```go
+type TaskResult struct {
+    Description string
+    Error error
+    RequeueAfter  time.Duration // Duration to requeue the task
+    Completed bool
+}
+
+// OperatorTask defines the interface for task execution.
+type TaskHandler interface {
+    EtcdReference() types.NamespacedName // based on if etcdReference is set in the spec.
+    Name() string
+    Type() druidv1alpha1.EtcdOperatorTaskType
+    Logger() logr.Logger
+    // Checks if the task is permitted to run. This is a one-time gate; once passed, it is not checked again for the same task execution.
+    Admit(ctx context.Context) *TaskResult  
+    Run(ctx context.Context) *TaskResult // The Run method will have to check if the necessary pre conditions hold true upon each call. 
+    Cleanup(ctx context.Context) *TaskResult // Will be triggered once the task is in a completed state.
+}
+
+```
+
+**Interface responsibilities:**
+
+- `Admit`: One-time precondition check before task execution.
+- `Run`: Main execution logic, called on each reconcile until completion.
+- `Cleanup`: Post-completion or TTL-based cleanup.
+- `Name`, `Type`, `EtcdReference`: Used for metrics and logging.
+
+
+## TaskHandler Instantiation
+
+The controller uses a factory function (`createTaskHandlerInstance`) to instantiate the correct handler for each task type. This enables easy extension and decouples task logic from the reconciler.
+
+```go
+// Add a new TaskHandler as shown below:
+func (r *Reconciler) createTaskHandlerInstance(task *v1alpha1.EtcdOperatorTask) (tasks.TaskHandler, error) {
+    switch task.Spec.Type {
+    case v1alpha1.EtcdOperatorTaskTypeOnDemandSnapshot:
+        return ondemandsnapshot.New(r.client, r.logger, task) // New() initializes the TaskHandler.
+    // Add more cases for other task types
+    default:
+        return nil, fmt.Errorf("unsupported task type: %s", task.Spec.Type)
+    }
+}
+
+```
+
+### Adding New Task Types
+
+To add a new task type:
+1. Implement the `TaskHandler` interface for your task.
+2. Register the handler in `createTaskHandlerInstance` with a new case for your type.
+
```

Comment:
How do you handle RBAC for tasks? Should there be granularity on who can execute what kind of operation task or we have a blanket RBAC for all tasks. 

Also shouldn't we audit trail any such task execution? I believe, we should have it included in the implementation where an execution to task handler is audit trailed?